components/downgrade_modal/index.tsx

```diff
@@ -0,0 +1,75 @@
+// Copyright (c) 2015-present Mattermost, Inc. All Rights Reserved.
+// See LICENSE.txt for license information.
+
+import React, {useEffect, useState} from 'react';
+import {useDispatch} from 'react-redux';
+
+import RootPortal from 'components/root_portal';
+import FullScreenModal from 'components/widgets/modals/full_screen_modal';
+import BackgroundSvg from 'components/common/svg_images_components/background_svg';
+
+import 'components/payment_form/payment_form.scss';
+
+import IconMessage from '../purchase_modal/icon_message';
+import {t} from 'utils/i18n';
+
+import './index.scss';
+
+import {closeModal} from 'actions/views/modals';
+import {ModalIdentifiers} from 'utils/constants';
+import GettingStartedSVG from 'components/common/svg_images_components/getting_started_svg';
+
+const MIN_PROCESSING_MILLISECONDS = 5000;
+const MAX_FAKE_PROGRESS = 95;
+
+export default function DowngradeModal() {
+    const modal = React.createRef();
+    const dispatch = useDispatch();
+    const [progress, setProgress] = useState(0);
+
+    const progressBar: JSX.Element | null = (
+        <div className='ProcessPayment-progress'>
+            <div
+                className='ProcessPayment-progress-fill'
+                style={{width: `${progress}%`}}
+            />
+        </div>
+    );
+
+    useEffect(() => {
+        const updateProgress = () => setProgress(progress + 1);
+        if (progress < MAX_FAKE_PROGRESS) {
+            setTimeout(updateProgress, MIN_PROCESSING_MILLISECONDS / MAX_FAKE_PROGRESS);
```

Comment:
Nit: We should add some code to not run `setProgress` if the component is unmounted. It's considered a bug/memory leak to try to set state on an unounted component, and it can happen that this modal gets closed by the component that controls its display while there is still a pending timeout that will happen.

One common way is to track the returned timeout id and clearTimeout when the component unmounts.  One example of this bookkeeping code is here: https://github.com/mattermost/mattermost-webapp/blob/master/components/preparing_workspace/preparing_workspace.tsx#L148-L158 . When you return a function from a useEffect callback, it is executed on component unmount, plus each time the effect is called again: https://beta.reactjs.org/apis/react/useEffect#parameters .

That would work ok here, but also might not be the best because this effect refires every time progress updates, so we could be calling clearTimeout several times. I don't think that would be harmful (because the clear timeout would always occur *after* the previous timeout got to run, since that is what drives the next timeout), it just seems excessive.

Another approach we could take is to add a mounted ref that is set to true initially. Then we have a separate useEffect with no dependencies that all it does is return a function that sets the ref value to false. Then, `updateProgress` can check the value in the mounted ref before calling setProgress.

Reply:
Ah, interesting. I thought the `if (progress < MAX_FAKE_PROGRESS) {` would catch this, but it makes sense that it doesn't since the timeout is still set prior to that condition becoming `false`.

I'll make an update to this, thanks for the examples :)